src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2626,163 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
 }
 FCIMPLEND
 
-// Return true if the valuetype does not contain pointer and is tightly packed
+typedef BOOL(*GetMethodTablePropertyFunc)(MethodTable*);
+typedef void(*SetMethodTablePropertyFunc)(MethodTable*);
+
+static BOOL HasOverriddenMethodWorker(MethodTable* mt, const char* methodName)
+{
+    MethodTable::IntroducedMethodIterator it(mt);
+    for (; it.IsValid(); it.Next())
+    {
+        if (it.GetMethodDesc()->IsVirtual()
+            && strcmp(it.GetMethodDesc()->GetName(), methodName) == 0)
+        {
+            return TRUE;
+        }
+    }    
+
+    return FALSE;
+}
+
+static BOOL HasOverriddenMethod(MethodTable* mt,
+                                const char* methodName,
+                                GetMethodTablePropertyFunc hasCheckedOverridden,
+                                SetMethodTablePropertyFunc setHasCheckedOverridden,
+                                GetMethodTablePropertyFunc hasOverridden,
+                                SetMethodTablePropertyFunc setHasOverridden)
+{
+    LIMITED_METHOD_CONTRACT
+
+    if (hasCheckedOverridden(mt))
+    {
+        return hasOverridden(mt);
+    }
+
+    BOOL hasOverriddenMethod = FALSE;
+
+    // Do we have the overriden method in current Type?
+    if (HasOverriddenMethodWorker(mt, methodName))
+    {
+        hasOverriddenMethod = TRUE;
+    }
+    else
+    {
+        // Now we need to do a depth-first-search on the instance fields.
+        Stack<ApproxFieldDescIterator> fieldIterStack;
+        fieldIterStack.push(ApproxFieldDescIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS));
+        while (!fieldIterStack.empty())
+        {
+            ApproxFieldDescIterator& iter = fieldIterStack.top();
+            FieldDesc* pField = iter.Next();
+            if (pField == NULL)
+            {
+                fieldIterStack.pop();
+                continue;
+            }
+
+            if (pField->GetFieldType() != ELEMENT_TYPE_VALUETYPE
+                && pField->GetFieldType() != ELEMENT_TYPE_CLASS)
+            {
+                // We cannot find overridden method in non-class types.
+                continue;
+            }
+
+            // Try to find the specified overriden method in current field type.
+            MethodTable* fieldMethodTable = pField->LookupFieldTypeHandle().GetMethodTable();
+            if (hasCheckedOverridden(fieldMethodTable))
+            {
+                // If we have checked this field type previously, it implies this field type
+                // does not have the specified overridden method.
+                continue;
+            }
+
+            setHasCheckedOverridden(fieldMethodTable);
+            if (HasOverriddenMethodWorker(fieldMethodTable, methodName))
```

Comment:
Consider the case when one thread calls `setHasCheckedOverridden`, then preempted before making the `HasOverriddenMethodWorker` call.  Another thread comes and calls `hasCheckedOverridden` on the same type.  That would incorrectly return `true`, even though we have not checked the type for the `Equals` override yet.